### PR TITLE
Fix docstring inheritance

### LINKED ISSUE #3033: _inherit_docstrings does not wrap classmethods and own attributes
> Another problem with `doc_checker` is that for file `modin/pandas/accessor.py` from PR #3022 I get errors about functions `from_spmatrix` and `from_coo` although docstrings for these functions should be inherited.

The problem has to do with how `_inherit_docstrings` handle functions with `classmethod` decorator. cc @vnlitvinov

_Originally posted by @anmyachev in https://github.com/modin-project/modin/issues/2995#issuecomment-828629542_

### PR DESCRIPTION
**Pull request by vnlitvinov**
## What do these changes do?
Fix handling `@staticmethod` and `@classmethod` decorators, add tests for `@_inherit_docstrings` (finally 🤣, should have made those from the start).

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3033 
- [x] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

**Review by anmyachev — APPROVED** [review]
LGTM!

File: modin/test/test_utils.py
Comment (devin-petersohn) [inline modin/test/test_utils.py:18]: Can we specify in a comment that these are for testing only? Otherwise looks good to me.

Comment (vnlitvinov) [reply modin/test/test_utils.py:20]: added in ccc8a54

**Review by anmyachev — APPROVED** [review]
LGTM

**Comment by dchigarev** [comment]
@vnlitvinov it seems that this PR has added another bug into `_inherit_docstrings` related to inheriting properties docstrings. In the following example, you can see that adding the `_inherit_docstrings` decorator overwrites the property itself, not only the docs.
```python
from modin.utils import _inherit_docstrings

def property_builder(name):
    def getter(*args):
        """Getter doc-string."""
        print(f"{name} getter")

    return property(getter)

class Parent:
    my_prop = property_builder("Parent")

@_inherit_docstrings(Parent)
class Derived(Parent):
    my_prop = property_builder("Derived")

class DerivedNoDocs(Parent):
    my_prop = property_builder("Derived")

Parent().my_prop        # Parent getter
Derived().my_prop       # Parent getter
DerivedNoDocs().my_prop # Derived getter
```
P.S. this bug is not presented in the current master, so something in this PR has broken.

**Comment by vnlitvinov** [comment]
It is impossible to override a docstring of the property as it's a read-only field. `_inherit_docstring` in `master` is assigning the property attribute as it is in this PR :)
However it's working with properties in the wrong way, assigning a property based on its getter name (not property name).

Thanks for the example, I've add a test based on that and will fix it soon™.

File: modin/utils.py
Comment (dchigarev) [inline modin/utils.py:182]: I would highlight in the docs of `_inherit_docstrings` that it may overwrite docstrings of the base classes of the decorated object. It's obvious when looking into the code, but not obvious at all when you're reading the function's contract.

```python
from modin.utils import _inherit_docstrings

class DocsContainer:
    def method():
        """Doc-string."""
        pass

class Base:
    def method(): # Not documented method
        pass

@_inherit_docstrings(DocsContainer)
class Derived(Base):
    pass

print(Derived.method.__doc__) # Doc-string.
print(Base.method.__doc__)    # Doc-string.

```

Comment (vnlitvinov) [reply modin/utils.py:182]: Good point!

**Review by dchigarev — DISMISSED** [review]

**Comment by dchigarev** [comment]
@devin-petersohn @anmyachev any other comments?

**Review by anmyachev — DISMISSED** [review]

**Review by devin-petersohn — COMMENTED** [review]
LGTM, please rebase after #3001 was merged.

Thanks!

**Comment by vnlitvinov** [comment]
@devin-petersohn it seems that #3001 has accidentally took most of the meat of this PR with it... so now it's basically fixing what @dchigarev found in the last comment :)

**Review by devin-petersohn — APPROVED** [review]
Thanks @vnlitvinov, LGTM

**Comment by devin-petersohn** [comment]
Restarted CI

**Pull request merged**
